### PR TITLE
[FAS3] - Test Suite

### PR DESCRIPTION
**Pull request by skrzepto**
Focus of this testsuite is on functional testing.

Current output `25 passed, 5 skipped in 6.45 seconds `

The skipped tests are current bugs in the program that need to be fixed.

Current Limitations
* Only a local postgres db is used with user postgres db named `fas`
    * Be able to pass parameters via testsuite to use faitout, sqlite, local pgsql, etc...
* ....

-----
Squashed commits
* when generating fake data need to delete foreign key relations first
* fixing base test class to reuse db connection and reset the testapp from pyramid
* there are 5 total people in testsuite
* need to delete activity log to delete people first
* removing html string comparison as its not a very good test
* fixing string comparison
* removing unecessary code, removing trailing space
* adding pending account notification when loging in
* removed utc info from date having issues with time aware dates, hardcoded the status since sqlalchemy complains about it not being valid and not spitting out an int
* updating count for dump since it returns total of active users instead.
* url of the resource is now full path, split up string to make tests pass
* updating string comparison test to pass
* adding packages to install for test suite
* issue with registering an account now, skip tests for now so test suite can be reviewed
* This was causing a datetime comparison crash, need to investigate if this is important
* adding notes on setup up pgsql

### CODE REVIEW

File: development.ini [hunk 1]
```diff
@@ -67,7 +67,9 @@ user.security_change.timeout = 60
 ###
 # Database settings
 ###
-sqlalchemy.url = sqlite:///%(here)s/fas.sqlite
+# sqlalchemy.url = sqlite:///%(here)s/fas.sqlite
+# postgresql+psycopg2://user:password@host:port/dbname
+sqlalchemy.url = postgresql+psycopg2://postgres:fas@localhost:5432/fas
```
Comment (pypingou) [inline development.ini]: publishing the URL here is meant? Looks like it contains a password

File: fas/scripts/admin.py [hunk 2]
```diff
@@ -365,8 +365,16 @@ def main(argv=sys.argv):
...
                 print 'Cleaning up People data.'
-                DBSession.query(People).delete()
+                DBSession.query(AccountPermissions).delete()
+
+                DBSession.query(GroupMembership).delete()
                 DBSession.query(Groups).delete()
+                DBSession.query(GroupType).delete()
+
+                DBSession.query(PeopleAccountActivitiesLog).delete()
+                DBSession.query(People).delete()
```
Comment (pypingou) [inline fas/scripts/admin.py]: This looks like DB configuration error, do they have ``ondelete=updated``?

File: fas/security.py [hunk 3]
```diff
@@ -292,7 +292,7 @@ def process_login(request, person, password):
             # headers = remember(request, person.username)
             request.session.get_csrf_token()
 
-            notify(LoginSucceeded(request, person))
+            # notify(LoginSucceeded(request, person))
```
Comment (pypingou) [inline fas/security.py]: I guess this is the issue you reported on the other ticket about broken login, something @laxathom needs to look into :)

File: development.ini (hunk 1 above)
Comment (skrzepto) [inline development.ini]: Since FAS3 is geared to postgres I wanted to create a default login for development and running the test suite. Perhaps I could delete line 72 and ask the new developers to edit line 71 to their own configuration.

File: fas/scripts/admin.py (hunk 2 above)
Comment (skrzepto) [inline fas/scripts/admin.py]: Seems that the models don't have `ondelete=updated` I'll look into it

File: tests/__init__.py [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+              datetime(2013, 8, 28, 14, 46, 44),  # tzinfo=UTC),
+              datetime(2016, 2, 28, 14, 46, 44), ]  # tzinfo=UTC)]
+
+    status = [8, 1, 3]  # DISABLED, ACTIVE, PENDING
+    tokens = ['abcd', '1234', 'xyz']
+    group_list = [300, 302, 304]  # avengers, fantastic_four, x-men
+    membership_status = [1, 1, 2]  # Approved, Approved, Pending
+    membership_role = [4, 2, 1]  # Admin, Editor, User
+
+    """
```
Comment (skrzepto) [inline tests/__init__.py]: Delete this

File: tests/__init__.py [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+                          enabled_at_signup=True)
+    DBSession.add(la)
+
+
+from fas import main
+from webtest import TestApp
+
+
+def getDBPath():
+    if False:  # pytest_params == 'faitout':
```
Comment (skrzepto) [inline tests/__init__.py]: Fix get dbpath

File: development.ini (hunk 1 above)
Comment (pypingou) [inline development.ini]: I think I would let user edit the configuration line, maybe we could just make it a mix of 72 and 71:
``sqlalchemy.url = postgresql+psycopg2://user:password@host:port/dbname``

File: fas/security.py (hunk 3 above)
Comment (laxathom) [inline fas/security.py]: Can you point me to the related ticket then?

File: fas/scripts/admin.py (hunk 2 above)
Comment (skrzepto) [inline fas/scripts/admin.py]: Fixed in in https://github.com/fedora-infra/fas/pull/222

**Comment by skrzepto** [comment]
I have the test suite run in Vagrant in a separate table. It seems to be the cleanest solution to test with an actual pgsql db that doesn't require internet connection.

File: README.rst [hunk 6]
```diff
@@ -153,8 +153,32 @@ Initialize the database
 
 Run the test suite
 ------------------
+``pytest``
 
-``% python setup.py test``
+*Note: Only works with pgsql at the moment*
```
Comment (skrzepto) [inline README.rst]: This is the manual method of running the test suite need to clean this up to just `vagrant ssh -c 'pushd /vagrant/; find . -name \*.pyc -delete ; pytest'`

File: development.ini [hunk 7]
```diff
@@ -70,7 +70,8 @@ user.security_change.timeout = 60
 ###
 # Database settings
 ###
-sqlalchemy.url = sqlite:///%(here)s/fas.sqlite
```
Comment (skrzepto) [inline development.ini]: Undo this change

**Comment by skrzepto** [comment]
@laxathom @pypingou Any thoughts on the test suite?

**Comment by laxathom** [comment]
While I like the idea of having the test suite vagrand-ready, I tend to say we should also provide a way to perform it w/o it like if we need to run the test suite from a CI system, I'm not sure vagrant is supported yet in such env.

**Comment by skrzepto** [comment]
@laxathom the test suite is able to run locally but you need to create the postgres table for it and modify to the development.ini file to use pgsql. So for example how I have it so far is if in the development.ini file has a sqalchmey url use pgsql with table name fas. The test suite appends `test` to table name so user needs to create `fastest` table to use for the test suite. Thats the only limitation.

**Comment by laxathom** [comment]
hm...can we maybe instead use another db name with `_test` rather than appending word to `_test` to all tables?

**Comment by skrzepto** [comment]
@laxathom just noticed i used the wrong name. I meant database instead of table. Hopefully that clears somethings up

**Comment by skrzepto** [comment]
Anything else needed to clean the test suite up? @laxathom
